src/Controls/src/SourceGen/RouteIncrementalGenerator.cs

```diff
@@ -0,0 +1,564 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+using static Microsoft.Maui.Controls.SourceGen.Helper;
+
+namespace Microsoft.Maui.Controls.SourceGen
+{
+	[Generator(LanguageNames.CSharp)]
+	public partial class RouteIncrementalGenerator : IIncrementalGenerator
+	{
+		static readonly string Indent = new string('\x20', 8);
+		const string AutoGeneratedHeaderText = """
+			//------------------------------------------------------------------------------
+			// <auto-generated>
+			//     This code was generated by a .NET MAUI source generator.
+			//
+			//     Changes to this file may cause incorrect behavior and will be lost if
+			//     the code is regenerated.
+			// </auto-generated>
+			//------------------------------------------------------------------------------
+			""";
+
+		public void Initialize(IncrementalGeneratorInitializationContext context)
+		{
+			var classDeclarations = context.CompilationProvider.Combine(context.SyntaxProvider.CreateSyntaxProvider(
+					static (node, _) => node is ClassDeclarationSyntax,
+					static (ctx, _) => GetSemanticTargetForGeneration(ctx))
+				.Where(static x => x is not null)
+				.Collect());
+
+			context.RegisterSourceOutput(classDeclarations, (spc, source) => Execute(spc, source.Left, source.Right));
+		}
+
+		private static ClassDeclarationSyntax? GetSemanticTargetForGeneration(GeneratorSyntaxContext context)
+		{
+			if (context.Node is not ClassDeclarationSyntax cds)
+			{
+				return null;
+			}
+
+			if (cds.Identifier.ValueText == "MauiProgram")
+			{
+				foreach (var member in cds.Members)
+				{
+					if (member is MethodDeclarationSyntax { Identifier.ValueText: "CreateMauiApp" })
+					{
+						return cds;
+					}
+				}
+			}
+
+			var routeAttribute = cds.AttributeLists.SelectMany(x => x.Attributes)
+				.Where(a => a.Name.NormalizeWhitespace().ToFullString() == "Route")
+				.FirstOrDefault();
+
+			if (routeAttribute is not null)
+			{
+				return cds;
+			}
+
+			var serviceAttribute = cds.AttributeLists.SelectMany(x => x.Attributes)
+				.Where(a => a.Name.NormalizeWhitespace().ToFullString() == "MauiService")
+				.FirstOrDefault();
+
+			if (serviceAttribute is not null)
+			{
+				return cds;
+			}
+
+			if (cds.BaseList is not null)
+			{
+				foreach (var baseType in cds.BaseList.Types)
+				{
+					if (baseType.Type is IdentifierNameSyntax identifierName)
+					{
+						return identifierName.Identifier.ValueText switch
+						{
+							"Application" => cds,
+							"Shell" => cds,
+							_ => null,
+						};
+					}
+				}
+			}
+
+			return null;
+		}
+
+		private static void Execute(SourceProductionContext context, Compilation compilation, ImmutableArray<ClassDeclarationSyntax?> classes)
```

Comment:
Your IncrementalGenerator is broken. You can't pass the Roslyn types to the `Execute` method, they should be just in the queries pipelines. 
The reason is that Roslyn types can't be compared, so there's no way to cache, and the code will be generated at every file change (no matter what file changes).

You can see the Source Generators implemented on the CommunityToolkit.MVVM or CommunityToolkit.Maui as reference